### PR TITLE
feat(v2/storage-4): add storage mode response download modals

### PR DESCRIPTION
**Pull request by karrui**
## Problem

This PR adds storage mode response download modals.

TODOs to be done in future PRs (too keep this PR smaller):
- [ ] add summary modal when download has completed, or errored.

Related to #3129

## Solution

**Breaking Changes** 

- [x] No - this PR is backwards compatible  

**Features**:
- progress modal will show up 5 seconds after download (if download is not yet complete) for non-attachment downloads
- show confirmation modal on downloading responses with attachments, which will segue to progress modal
- allow cancellation of decryption
- add prompt when user attempts to leave the page
- add Progress component theme

## Before & After Screenshots
New stories have been added for `DownloadWithAttachmentModal` and `ProgressModal` component.

### CODE REVIEW

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx [hunk 1]
```diff
@@ -1,29 +1,62 @@
...
+  const { downloadParams, responsesCount } = useStorageResponsesContext()
+  const [_downloadCount, setDownloadCount] = useState(0)
+  const downloadCount = useThrottle(_downloadCount, 1000)
+
+  const downloadPercentage = useMemo(() => {
+    if (!responsesCount) return 0
+    return Math.floor((downloadCount / responsesCount) * 100)
+  }, [downloadCount, responsesCount])
+
+  useTimeout(onProgressModalOpen, progressModalTimeout)
```
Comment (mantariksh) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx:42]: so the timeout is initially `null` which doesn't do anything, and then the timeout gets set when `setProgressModalTimeout` is called and hence `progressModalTimeout` changes?

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx [hunk 2]
```diff
@@ -38,35 +71,72 @@ export const DownloadButton = (): JSX.Element => {
...
-              </Menu.Item>
-            </Menu.List>
-          </>
-        )}
-      </Menu>
-    </Box>
+    <>
+      {responsesCount && (
+        <DownloadWithAttachmentModal
+          responsesCount={responsesCount}
```
Comment (mantariksh) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx:89]: nit: feel like we're passing this prop everywhere when we can just use `useStorageResponsesContext`

**Review by mantariksh — APPROVED** [review]
tbh getting increasingly hard to review in detail without writing code but generally lgtm

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadWithAttachmentModal/DownloadWithAttachmentModal.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  onCancel: () => void
+  isDownloading: boolean
+  responsesCount: number
+  downloadPercentage: number
+}
+
+export enum DownloadWithAttachmentFlowStates {
+  Confirmation = 'confirmation',
+  Progress = 'progress',
+  Complete = 'complete',
```
Comment (hanstirtaputra) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadWithAttachmentModal/DownloadWithAttachmentModal.tsx:28]: When would be using the `DownloadWithAttachmentFlowStates.Complete` state?

Comment (hanstirtaputra) [inline frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadWithAttachmentModal/DownloadWithAttachmentModal.tsx:28]: When would be using the `DownloadWithAttachmentFlowStates.Complete` state?

**Review by hanstirtaputra — COMMENTED** [review]
Tested locally for downloading responses with attachments. Downloading the same responses in "CSV with attachments" results in the second download not progressing. Is this intended?

https://user-images.githubusercontent.com/24553546/173756277-311342e3-01a1-461c-aac2-944cc196431e.mov

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadWithAttachmentModal/DownloadWithAttachmentModal.tsx (hunk 3 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadWithAttachmentModal/DownloadWithAttachmentModal.tsx:28]: coming in next pr, for download completion summary.

**Comment by karrui** [comment]
> Tested locally for downloading responses with attachments. Downloading the same responses in "CSV with attachments" results in the second download not progressing. Is this intended?
> 
>  test-download-with-attachments.mov

Thanks for the catch! Will fix.

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx (hunk 2 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx:89]: wanted to keep the modal "pure" for storybook, which i have now added stories for hehe

File: frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx (hunk 1 above)
Comment (karrui) [reply frontend/src/features/admin-form/responses/ResponsesPage/storage/UnlockedResponses/DownloadButton.tsx:42]: yes

File: frontend/src/hooks/useBlocker.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+interface Navigator extends BaseNavigator {
+  block: History['block']
+}
+
+type NavigationContextWithBlock = ContextType<typeof NavigationContext> & {
+  navigator: Navigator
+}
+
+/**
+ * @source https://github.com/remix-run/react-router/commit/256cad70d3fd4500b1abcfea66f3ee622fb90874
```
[the remark this answers is not in the document]
Comment (karrui) [reply frontend/src/hooks/useBlocker.ts:22]: added

**Pull request merged**